tools/check_wav_file_snr.m

```diff
@@ -0,0 +1,439 @@
+		fprintf(1, 'Error: Channel %d has no signal\n', ch);
+		error('Failed.');
+	end
+
+	% STFT
+	[stft, param] = compute_stft(x(:, ch), fs, param);
+
+	% Check frequency, get STFT frequency index
+	[signal_idx, tone_f] = find_test_tone(stft, param, test_tone_f);
+
+	% Get levels and SNR estimates
+	meas = get_tone_levels(stft, param, signal_idx);
+
+	% Checks for levels data from FFTs, after stable level
+	% and time to ignore
+	meas = check_tone_levels(stft, param, meas);
+
+	% Check begin of wav file, before stable level
+	meas = check_time_domain(x(:,ch), param, meas);
+
+	% If poor SNR check if there's periodic glitches
+	meas = check_glitch_periodicity(x(:,ch), param, meas, ch);
+
+	% Plot
+	fnstr = sprintf('%s_ch%d', fn, ch);
+	idstr = sprintf('%s ch%d', fn, ch);
+	if meas.num_glitches > 0
+		plot_glitch(param, x(:, ch), meas, idstr, fnstr);
+	end
+
+	all_stft(ch, :, :) = stft;
+	all_meas(ch) = meas;
+	if meas.success == false
+		test_passed = false;
+	end
+end
+
+plot_specgram(param, all_stft, fn, channels);
+plot_levels(param, all_meas, fn, channels);
+
+
+fprintf(1, 'Ch, Pass, Frequency, SNR min, SNR avg, Signal avg, Noise avg, Noise max, Num glitch, 1st glitch\n');
+
+for ch = 1 : channels
+	meas = all_meas(ch);
+	fprintf(1, '%d, %2d,     %5.0f,  %6.1f,  %6.1f,     %6.1f,    %6.1f,    %6.1f,     %6d,  %8.3f\n', ...
+		ch, meas.success, tone_f, meas.min_snr_db, meas.mean_snr_db, ...
+		meas.mean_signal_db, meas.mean_noise_db, ...
+		meas.max_noise_db, meas.num_glitches, meas.t_glitch);
+end
+
+if test_passed
+	fprintf(1, 'Passed\n');
+else
+	error('Failed');
+end
+
+end
+
+%
+% Helper functions
+%
+
+function meas = check_time_domain(x, param, meas)
+
+% High-pass filter to find glitches
+idx_end = round(1 + meas.t_start * param.fs);
+[b, a] = butter(2, 0.5, 'high');
+y = x(1 : idx_end);
+z = abs(filter(b, a, y));
+
+% Skip start transient
+s = max(abs(y));
+idx1 = find(z < 0.01 * s, 1, 'first');
+idx2 = find(z(idx1 : end) > 0.02 * s);
+
+if ~isempty(idx2)
+	meas.num_glitches = meas.num_glitches + 1;
+	meas.t_glitch = (idx1 + idx2(1) - 1) / param.fs;
+end
+
+end
+
+function meas = check_glitch_periodicity(x, param, meas, ch)
+
+if meas.success
+	return
+end
+
+i1 = round(param.fs * param.t_ignore_from_start);
+i2 = length(x) - round(param.fs * param.t_ignore_from_end);
+[b, a] = butter(2, 0.95, 'high');
+y = abs(filter(b, a, x));
+z = y(i1:i2);
+thr = mean(z) + std(z);
+[~, locs]= findpeaks(z, 'MinPeakHeight', thr);
+dlocs = diff(locs);
+if isempty(dlocs)
+	return
+end
+
+[counts, centers] = hist(dlocs);
+[counts, i] = sort(counts, 'descend');
+centers = centers(i) * 1e3 / param.fs;
+thr = mean(counts) + std(counts);
+idx = find(counts > thr);
+if ~isempty(idx)
+	fprintf(1, 'Ch%d periodic glitches possibly every', ch)
+	for i = 1 : length(idx)
+		fprintf(' %.1f ms (n = %d)', centers(idx(i)), counts(idx(i)));
+	end
+	fprintf(1, '\n');
+end
+
+end
+
+function success = signal_found(x)
+
+% All zeros or DC
+if abs(min(x) - max(x)) < eps
```

Comment:
I added this to fail if signal is all digital zeros. Eps should work in all cases to detect such. But I'll check that it works in both.

Reply:
please refer the link, 
Matlab : [https://in.mathworks.com/help/matlab/ref/eps.html](url)
Octave:  [https://octave.sourceforge.io/octave/function/eps.html](url)
check works if `x` data limits are in bounds .

I'd like to know why `min(x)-max(x)` is taken into consideration rather than `max(x)-min.(x)`,

Reply:
On my computer both Matlab and Octave return the same eps

```
>> eps

ans =

   2.2204e-16
```
Do you have a suggestion how this line should be fixed to avoid the concern that it doesn't work? 10 * eps should be fine as well?

Reply:
OK, I see the following information, and because the constant is small, I believe there is a low possibility of failure.
K>> sprintf('matlab =  %.32f, fixedOctave = %.32f',eps,2.2204e-16)

ans =

    'matlab =  0.00000000000000022204460492503131, fixedOctave = 0.00000000000000022204000000000000'

Reply:
The notes are an approximation. Octave should still return the value returned by matlab (which is a negative power of 2, being the ULP for double precision 1.0), it would be extremely weird if it didn't.

So there should be no concern about how it works or doesn't work. I don't know where the myth that the value is treated differently between Matlab and Octave comes from (shouldn't the value of epsilon come from the C standard library anyway for both? Epsilon is defined as the ULP for 1.0, that is the difference between 1.0 and the smallest double precision value greater than 1.0, which is 2^-52 indeed on any platform with the usual double precision floats)

2.2204e-16 is an approximation to the correct epsilon. You can use it if you want to, but there's no real benefit to that over the actual epsilon value.

Reply:
I double-checked the `eps` findings in Matlab and Octave, and they are identical [`0.0000000000000002220446049250313080847263336181640625`]. Please disregard the comment. Thank you very much.